force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,15 +218,37 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
 
-		// Extract the Account Name from Contact's LastName
+        Set<String> lastNames = new Set<String>();
+        // Loop through each Contact
+        for (Contact contact : contactsToUpsert) lastNames.add(contact.LastName);
+
+        Map<String, Account> accountsWithLastName = new Map<String, Account>();
+        List<Account> accountsToInsert = new List<Account>();
```

Comment:
Excellent, a lot of people didn't get this approach. Great work with this question.